# Aula 06 - Exercício 05

### Alunos:
 - Alexandre Batistella Bellas, 9763168
 - Moisés Botarro Ferraz Silva, 8504135

## Multilayer Perceptron para classificação do dataset Wine
Neste exercício utilizaremos a base de dados Wine, que consiste em 178 exemplos de dimensionalidade 13, separados em 3 classes. Os atributos dessa base de dados não estão normalizados, apresentando intervalos variados de valores.

- Carregue a base de dados wine e normalize entre 0 e 1 o valor dos seus atributos

In [1]:
from sklearn.datasets import load_wine
from sklearn.preprocessing import MinMaxScaler

data = load_wine()
scaler = MinMaxScaler(feature_range=(0,1))

print("--- original values ---\n%s" % data.data)
data.data = scaler.fit_transform(data.data)
print("--- normalized values ---\n%s" % data.data)

--- original values ---
[[1.423e+01 1.710e+00 2.430e+00 ... 1.040e+00 3.920e+00 1.065e+03]
 [1.320e+01 1.780e+00 2.140e+00 ... 1.050e+00 3.400e+00 1.050e+03]
 [1.316e+01 2.360e+00 2.670e+00 ... 1.030e+00 3.170e+00 1.185e+03]
 ...
 [1.327e+01 4.280e+00 2.260e+00 ... 5.900e-01 1.560e+00 8.350e+02]
 [1.317e+01 2.590e+00 2.370e+00 ... 6.000e-01 1.620e+00 8.400e+02]
 [1.413e+01 4.100e+00 2.740e+00 ... 6.100e-01 1.600e+00 5.600e+02]]
--- normalized values ---
[[0.84210526 0.1916996  0.57219251 ... 0.45528455 0.97069597 0.56134094]
 [0.57105263 0.2055336  0.4171123  ... 0.46341463 0.78021978 0.55064194]
 [0.56052632 0.3201581  0.70053476 ... 0.44715447 0.6959707  0.64693295]
 ...
 [0.58947368 0.69960474 0.48128342 ... 0.08943089 0.10622711 0.39728959]
 [0.56315789 0.36561265 0.54010695 ... 0.09756098 0.12820513 0.40085592]
 [0.81578947 0.66403162 0.73796791 ... 0.10569106 0.12087912 0.20114123]]


---
- Agora divida a base em conjunto de teste e treino. Utilize 20% da base para teste e 80% para treino.

In [2]:
from sklearn.model_selection import train_test_split
import random 

random.seed(0)

x_train, x_test, y_train, y_test = train_test_split(data.data, data.target, test_size=0.2, shuffle=True, random_state=10)
print("size of training set: %s" % len(x_train))
print("size of testing set: %s" % len(x_test))
print("%.2f%% of data are used for test" % (len(x_test)*100/(len(data.data))))

size of training set: 142
size of testing set: 36
20.22% of data are used for test


---
- Treine 3 classificadores MLP (com `solver='sgd'`), variando o número de neurônios na camada escondida. Reporte o score dos classificadores nos conjuntos de treino e teste.
  - Parâmetros para  alterar: hidden_layer_sizes=(25,), max_iter=1000,
                      solver='sgd', tol=1e-4, random_state=1,
                      learning_rate_init=.1

In [3]:
from sklearn.neural_network import MLPClassifier

mlp1 = MLPClassifier(hidden_layer_sizes=(1,),solver='sgd', tol=1e-4, random_state=1, learning_rate_init=.1)
mlp1.fit(x_train, y_train)

print("--- 1 neuron in the hidden layer ---")
print("predicted %s " % mlp1.predict(x_test))
print("wanted    %s " % y_test)
print("accuracy: %s" % mlp1.score(x_test,y_test))

--- 1 neuron in the hidden layer ---
predicted [1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1] 
wanted    [1 1 0 1 0 1 1 0 2 0 0 1 0 1 1 1 1 1 1 2 0 2 0 0 1 2 1 2 1 1 2 2 1 2 1 0] 
accuracy: 0.5


In [4]:
mlp3 = MLPClassifier(hidden_layer_sizes=(3,),solver='sgd', tol=1e-4, random_state=1, learning_rate_init=.1)
mlp3.fit(x_train, y_train)

print("--- 3 neuron in the hidden layer ---")
print("predicted %s " % mlp3.predict(x_test))
print("wanted    %s " % y_test)
print("accuracy: %s" % mlp3.score(x_test,y_test))

--- 3 neuron in the hidden layer ---
predicted [1 1 0 1 0 1 1 0 2 0 0 1 0 2 1 1 2 1 1 2 0 2 0 0 1 2 2 2 1 1 2 2 1 2 1 0] 
wanted    [1 1 0 1 0 1 1 0 2 0 0 1 0 1 1 1 1 1 1 2 0 2 0 0 1 2 1 2 1 1 2 2 1 2 1 0] 
accuracy: 0.9166666666666666


In [5]:
mlp10 = MLPClassifier(hidden_layer_sizes=(10,),solver='sgd', tol=1e-4, random_state=1, learning_rate_init=.1)
mlp10.fit(x_train, y_train)

print("--- 10 neuron in the hidden layer ---")
print("predicted %s " % mlp10.predict(x_test))
print("wanted    %s " % y_test)
print("accuracy: %s" % mlp10.score(x_test,y_test))

--- 10 neuron in the hidden layer ---
predicted [1 1 0 1 0 1 1 0 2 0 0 1 0 1 1 1 2 1 1 2 0 2 0 0 1 2 0 2 1 1 2 2 1 2 1 0] 
wanted    [1 1 0 1 0 1 1 0 2 0 0 1 0 1 1 1 1 1 1 2 0 2 0 0 1 2 1 2 1 1 2 2 1 2 1 0] 
accuracy: 0.9444444444444444


---
- Selecione uma amostra de tamanho 10 do conjunto de testes. Selecione o classificador que apresentou maior acurácia no conjunto de testes e calcule, para cada elemento da amostra, a classe esperada, a classe obtida e a probabilidade estimada de cada classe

O classificador que apresentou melhor acurácia foi aquele com 10 neurônios na camada intermediária.

In [6]:
# Vamos pegar um conjunto de amostras de tamanho 10 que inclua o exemplo classificado erroneamento para mpl10.

x_test_10 = x_test[10:20]
y_test_10 = y_test[10:20]

predicted = mlp10.predict(x_test_10)
prob = mlp10.predict_proba(x_test_10)

print("wanted:    %s" % y_test_10)
print("predicted: %s" % predicted)
print("accuracy:  %s\n" % mlp10.score(x_test_10, y_test_10))

print("--- predicted probabilities ---")
print("ID \tWANTED \tPREDICT\tCLASS 0 \tCLASS 1\t \tCLASS 2")
for (el,score) in zip(range(0,len(prob)),prob):
    print("%i \t%i \t%i \t%.10f\t%.10f\t%.10f" % (el, y_test_10[el], predicted[el], score[0], score[1], score[2]))

wanted:    [0 1 0 1 1 1 1 1 1 2]
predicted: [0 1 0 1 1 1 2 1 1 2]
accuracy:  0.9

--- predicted probabilities ---
ID 	WANTED 	PREDICT	CLASS 0 	CLASS 1	 	CLASS 2
0 	0 	0 	0.9997354555	0.0002263466	0.0000381979
1 	1 	1 	0.0097434700	0.9902552668	0.0000012631
2 	0 	0 	0.9993213421	0.0006094791	0.0000691789
3 	1 	1 	0.0074663873	0.6566399478	0.3358936649
4 	1 	1 	0.0127505262	0.9871943412	0.0000551326
5 	1 	1 	0.0008270471	0.9990690221	0.0001039308
6 	1 	2 	0.0028499246	0.2378374667	0.7593126087
7 	1 	1 	0.0148901441	0.9849755791	0.0001342767
8 	1 	1 	0.0000668693	0.9999330333	0.0000000974
9 	2 	2 	0.0018514388	0.0003945606	0.9977540006


---
- Agora sobre a mesma amostra escolhida anteriormente, selecione o classificador que apresentou menor acurácia no conjunto de testes e calcule, para cada elemento da amostra, a classe esperada, a classe obtida e a probabilidade esperada de cada classe.

In [7]:
predicted = mlp1.predict(x_test_10)
prob = mlp1.predict_proba(x_test_10)

print("wanted:    %s" % y_test_10)
print("predicted: %s" % predicted)
print("accuracy:  %s\n" % mlp1.score(x_test_10, y_test_10))

print("--- predicted probabilities ---")
print("ID \tWANTED \tPREDICT\tCLASS 0 \tCLASS 1\t \tCLASS 2")
for (el,score) in zip(range(0,len(prob)),prob):
    print("%i \t%i \t%i \t%.10f\t%.10f\t%.10f" % (el, y_test_10[el], predicted[el], score[0], score[1], score[2]))

wanted:    [0 1 0 1 1 1 1 1 1 2]
predicted: [1 1 1 1 1 1 1 1 1 1]
accuracy:  0.7

--- predicted probabilities ---
ID 	WANTED 	PREDICT	CLASS 0 	CLASS 1	 	CLASS 2
0 	0 	1 	0.3042266611	0.4074168246	0.2883565143
1 	1 	1 	0.3042266611	0.4074168246	0.2883565143
2 	0 	1 	0.3042266611	0.4074168246	0.2883565143
3 	1 	1 	0.3042266611	0.4074168246	0.2883565143
4 	1 	1 	0.3042266611	0.4074168246	0.2883565143
5 	1 	1 	0.2805607042	0.4304869619	0.2889523339
6 	1 	1 	0.3042266611	0.4074168246	0.2883565143
7 	1 	1 	0.3042266611	0.4074168246	0.2883565143
8 	1 	1 	0.3042266611	0.4074168246	0.2883565143
9 	2 	1 	0.3042266611	0.4074168246	0.2883565143




---

- Faça uma análise das probabilidades calculadas por cada classificador nos exemplos corretos e relacione-as com a 'qualidade' dos mesmos.

No caso do classificador com 10 neurônios na camada intermediária, a maioria dos exemplos foi classificado na classe correta com uma certeza acima de 98%. A exceção está na amostra 3 que, apesar de ser classificada corretamente, houve uma maior incerteza na classificação, refletida pelos valores de probabilidade (65.66% para classe 1 x 33.59% para classe 2). Uma amostra (id 6) foi classificada erroneamente mas também com uma maior incerteza associada (75.93% ao invés dos altos valores para as demais amostras).

Para o classificador com apenas 1 neurônio na camada intermediária, 3 amostras foram classificadas de maneira errada. Entretanto, mesmo as classificadas corretamente apresentam um elevado grau de incerteza na classificação. Repare que as probabilidades de pertencer às classes 0, 1 e 2 são basicamente as mesmas para todos os exemplos, o que indica que apenas um neurônio não foi capaz de capturar informações suficientes para diferenciar as classes.

Podemos concluir, dessa forma, que mesmo para as amostras classificadas corretamente pelos dois modelos, o com 10 neurônios possui uma maior certeza nessa classificação.

---
- Calcule, usando 10-fold cross-validation, a acurácia média da melhor configuração de classificador que você utilizou e seu desvio padrão

In [8]:
from sklearn.model_selection import StratifiedKFold
import numpy as np

skf = StratifiedKFold(n_splits=10)

X = data.data
Y = data.target
acs = []

for train_index, test_index in skf.split(X,Y):
    x_train, x_test = X[train_index], X[test_index]
    y_train, y_test = Y[train_index], Y[test_index]
    acs.append(mlp10.score(x_test, y_test))
    
mean, std = np.mean(acs), np.std(acs)
print("accuracies: %s" % acs)
print("mean:       %s" % mean)
print("std :       %s" % std)

accuracies: [1.0, 0.9444444444444444, 1.0, 0.9444444444444444, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]
mean:       0.9888888888888889
std :       0.022222222222222233


Observando os valores de acurácia, é possível concluir que a cross validation é um melhor método para avaliar a performance do nosso modelo. Quando treinamos o MLP para 10 neurônios, usamos apenas um conjunto de teste. Caso esse conjunto de teste fosse um dos quais a acurácia é igual à 1, poderíamos achar que nosso modelo estava com alta performance. Entretanto, utilizando outros elementos para o conjunto de testes, chegaríamos à uma acurácia menor. 

Aplicando a Stratified K Fold, obtemos uma avaliação mais precisa para o nosso modelo uma vez que variamos o nosso conjunto de testes, o que permite treinar melhor o modelo e chegar a um resultado final mais assertivo. 